pkg/kubectl/resource_filter.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// FilterFunc is a function that knows how to filter a specific resource kind.
+// It receives a generic runtime.Object which must be type-checked by the function.
+// Returns a boolean value true if a resource is filtered, or false otherwise.
+type FilterFunc func(runtime.Object, PrintOptions) bool
+
+// ResourceFilter filters api resources
+type ResourceFilter interface {
+	Filter(obj runtime.Object) (bool, error)
+	AddFilter(reflect.Type, FilterFunc) error
```

Comment:
I don't see the need for this abstraction.  It would be better if you simply accepted a list of functions, and functions did type casts to see if they recognized the type.  Reflect and maps are very slow - a simple `if obj, ok := input.(MyType); ok` will be 2-3 orders of magnitude faster and avoid the need for reflection.

If you find yourself needing to change the list of filters, just wrap them.  You shouldn't need to have a mutator on the filter in order to be able to accomplish your goal:

```
type Filters []Filterer

func (f Filters) Filter(obj runtime.Object) (bool, error) {
  for _, filter := range f {
    ok, err := filter.Filter(obj)
    if err != nil {
      return false, err
    }
    if ok {
      return true, nil
    }
  }
  return false, nil
}
```

Is much much simpler.  Filter count accumulation doesn't belong in this interface - it belongs in the thing calling Filter.